pkg/kubelet/kubelet.go

```diff
@@ -2067,11 +2067,14 @@ func (kl *Kubelet) SyncTerminatingPod(_ context.Context, pod *v1.Pod, podStatus
 	// catch race conditions introduced by callers updating pod status out of order.
 	// TODO: have KillPod return the terminal status of stopped containers and write that into the
 	//  cache immediately
-	podStatus, err := kl.containerRuntime.GetPodStatus(ctx, pod.UID, pod.Name, pod.Namespace)
+	stoppedPodStatus, err := kl.containerRuntime.GetPodStatus(ctx, pod.UID, pod.Name, pod.Namespace)
 	if err != nil {
 		klog.ErrorS(err, "Unable to read pod status prior to final pod termination", "pod", klog.KObj(pod), "podUID", pod.UID)
 		return err
 	}
+	// Copy the IPs from the previous status, because containerRuntime.GetPodStatus
```

Comment:
I wonder if kubelet restart in a very precise moment will break this? Even if so - the chance is slim. I just wonder if this was a problem before this regression?

Reply:
To avoid that, we might want to update [`mergePodStatus`](https://github.com/kubernetes/kubernetes/blob/8c508c5480844682eab8a4172526b6db3476fdc0/pkg/kubelet/status/status_manager.go#L1049) to reuse the IPs from the `oldPodStatus` if:
1. `newPodStatus` is terminated
2. `newPodStatus` IPs are empty

Then we can at least make sure we're never overwriting the IPs stored in the API after a pod has terminated.

Reply:
Can I ask a very general question?

When is the podIP "released" by the Kubelet?  

Is it when the sandbox is stopped (https://github.com/kubernetes/kubernetes/blob/9895910846eeccf0694c577ac439f492ae2dfc3c/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L1374-L1375)?  Or is it later?

I'm trying to determine the earliest point during termination at which a pod is considered to not "own" that podIP, such that CNI might give out that IP to another pod (not a specific CNI provider, but any CNI provider). Once CNI might give a terminal pod's IP out to another pod is when subsequent apiserver pod status updates should no longer contain that podIP - preserving the podIP in status would *not* be desirable.

So if killPod ensures that a podIP is "released" from the perspective of CNI, then when this function exits we statusManager should consider the pod as having no IP since it is now officially "terminal" (exiting SyncTerminatingPod is a state transition for the pod worker).

Reply:
I don't think we should remove the PodIP from status, I'm trying to find more information from previous conversations,  but I think we concluded the expectations are that the pods keep the IP reported https://github.com/kubernetes/kubernetes/issues/109414#issuecomment-1126203054

A IP reported multiple times is not a problem, is only a problem if there are multiple Pods that are active with the same IP

Reply:
> I wonder if kubelet restart in a very precise moment will break this? 

I tested kubelet restart by inducing kubelet crash with a snippet like this:
```golang
   if strings.HasPrefix(pod.Name, "pod-") {
       markerFile := "/tmp/restarted"
       _, err := os.Stat(markerFile)
       if os.IsNotExist(err) {
           if file, err := os.Create(markerFile); err != nil {
               file.Close()
           }
           os.Exit(1)
       } else {
           if err := os.Remove(markerFile); err != nil {
               klog.ErrorS(err, "MYDEBUGError removign marker file", "pod", klog.KObj(pod), "podUID", pod.UID)
           }
       }
   }
```
And indeed the `PodIP` is lost if I inject it between [after killed](https://github.com/kubernetes/kubernetes/blob/e832b7023066a6f386634ad5d99e574281e1e6b4/pkg/kubelet/kubelet.go#L2064) and [before TerminatePod](https://github.com/kubernetes/kubernetes/blob/e832b7023066a6f386634ad5d99e574281e1e6b4/pkg/kubelet/kubelet.go#L2228) (it would not fail 100% of times though). 

> Even if so - the chance is slim. I just wonder if this was a problem before this regression?

Yes, I also tested on 1.26 by injecting the same snippet, and it also fails, I think with a slightly more frequently, but don't have enough repeats to be definitive about the frequency.

Reply:
> To avoid that, we might want to update [`mergePodStatus`](https://github.com/kubernetes/kubernetes/blob/8c508c5480844682eab8a4172526b6db3476fdc0/pkg/kubelet/status/status_manager.go#L1049) to reuse the IPs from the `oldPodStatus` if:
> 
> 1. `newPodStatus` is terminated
> 2. `newPodStatus` IPs are empty
> 
> Then we can at least make sure we're never overwriting the IPs stored in the API after a pod has terminated.

This does not work by itself, because in that case we have one interim update when phase=Running, which clears the IP at the API server, here is the example watch output:

```
> kubectl get pods -w --output-watch-events -ocustom-columns=EVENT:.type,NAME:.object.metadata.name,IP:.object.status.podIP,PHASE:.object.status.phase,STATUS:.object.status | ts "%Y-%m-%d %H:%M:%.S"
2024-06-28 13:46:42.455092 EVENT   NAME           IP       PHASE     STATUS
2024-06-28 13:46:42.455310 ADDED   pod-pi-f4v4d   <none>   Pending   map[phase:Pending qosClass:BestEffort]
2024-06-28 13:46:42.459023 MODIFIED   pod-pi-f4v4d   <none>   Pending   map[conditions:[map[lastProbeTime:<nil> lastTransitionTime:2024-06-28T11:46:42Z status:True type:PodScheduled]] phase:Pending qosClass:BestEffort]
2024-06-28 13:46:42.469784 MODIFIED   pod-pi-f4v4d   <none>   Pending   map[conditions:[map[lastProbeTime:<nil> lastTransitionTime:2024-06-28T11:46:42Z status:False type:PodReadyToStartContainers] map[lastProbeTime:<nil> lastTransitionTime:2024-06-28T11:46:42Z status:True type:Initialized] map[lastProbeTime:<nil> lastTransitionTime:2024-06-28T11:46:42Z message:containers with unready status: [pi] reason:ContainersNotReady status:False type:Ready] map[lastProbeTime:<nil> lastTransitionTime:2024-06-28T11:46:42Z message:containers with unready status: [pi] reason:ContainersNotReady status:False type:ContainersReady] map[lastProbeTime:<nil> lastTransitionTime:2024-06-28T11:46:42Z status:True type:PodScheduled]] containerStatuses:[map[image:howardjohn/shell imageID: lastState:map[] name:pi ready:false restartCount:0 started:false state:map[waiting:map[reason:ContainerCreating]] volumeMounts:[map[mountPath:/var/run/secrets/kubernetes.io/serviceaccount name:kube-api-access-5nnd6 readOnly:true recursiveReadOnly:Disabled]]]] hostIP:192.168.8.2 hostIPs:[map[ip:192.168.8.2]] phase:Pending qosClass:BestEffort startTime:2024-06-28T11:46:42Z]
2024-06-28 13:46:46.101598 MODIFIED   pod-pi-f4v4d   10.244.1.2   Running   map[conditions:[map[lastProbeTime:<nil> lastTransitionTime:2024-06-28T11:46:46Z status:True type:PodReadyToStartContainers] map[lastProbeTime:<nil> lastTransitionTime:2024-06-28T11:46:42Z status:True type:Initialized] map[lastProbeTime:<nil> lastTransitionTime:2024-06-28T11:46:46Z status:True type:Ready] map[lastProbeTime:<nil> lastTransitionTime:2024-06-28T11:46:46Z status:True type:ContainersReady] map[lastProbeTime:<nil> lastTransitionTime:2024-06-28T11:46:42Z status:True type:PodScheduled]] containerStatuses:[map[containerID:containerd://7451597e9b3a1f9705ab6c87e3de2534f19ac89d21066f1c971b32b768932988 image:docker.io/howardjohn/shell:latest imageID:docker.io/howardjohn/shell@sha256:23546d7e285397394c46f44ecb503309096a24724cd1b20b841dfb49d8d6fb6f lastState:map[] name:pi ready:true restartCount:0 started:true state:map[running:map[startedAt:2024-06-28T11:46:45Z]] volumeMounts:[map[mountPath:/var/run/secrets/kubernetes.io/serviceaccount name:kube-api-access-5nnd6 readOnly:true recursiveReadOnly:Disabled]]]] hostIP:192.168.8.2 hostIPs:[map[ip:192.168.8.2]] phase:Running podIP:10.244.1.2 podIPs:[map[ip:10.244.1.2]] qosClass:BestEffort startTime:2024-06-28T11:46:42Z]
2024-06-28 13:46:48.106098 MODIFIED   pod-pi-f4v4d   10.244.1.2   Running   map[conditions:[map[lastProbeTime:<nil> lastTransitionTime:2024-06-28T11:46:46Z status:True type:PodReadyToStartContainers] map[lastProbeTime:<nil> lastTransitionTime:2024-06-28T11:46:42Z reason:PodCompleted status:True type:Initialized] map[lastProbeTime:<nil> lastTransitionTime:2024-06-28T11:46:48Z reason:PodCompleted status:False type:Ready] map[lastProbeTime:<nil> lastTransitionTime:2024-06-28T11:46:48Z reason:PodCompleted status:False type:ContainersReady] map[lastProbeTime:<nil> lastTransitionTime:2024-06-28T11:46:42Z status:True type:PodScheduled]] containerStatuses:[map[containerID:containerd://7451597e9b3a1f9705ab6c87e3de2534f19ac89d21066f1c971b32b768932988 image:docker.io/howardjohn/shell:latest imageID:docker.io/howardjohn/shell@sha256:23546d7e285397394c46f44ecb503309096a24724cd1b20b841dfb49d8d6fb6f lastState:map[] name:pi ready:false restartCount:0 started:false state:map[terminated:map[containerID:containerd://7451597e9b3a1f9705ab6c87e3de2534f19ac89d21066f1c971b32b768932988 exitCode:0 finishedAt:2024-06-28T11:46:47Z reason:Completed startedAt:2024-06-28T11:46:45Z]] volumeMounts:[map[mountPath:/var/run/secrets/kubernetes.io/serviceaccount name:kube-api-access-5nnd6 readOnly:true recursiveReadOnly:Disabled]]]] hostIP:192.168.8.2 hostIPs:[map[ip:192.168.8.2]] phase:Running podIP:10.244.1.2 podIPs:[map[ip:10.244.1.2]] qosClass:BestEffort startTime:2024-06-28T11:46:42Z]
2024-06-28 13:46:51.658728 MODIFIED   pod-pi-f4v4d   <none>       Running   map[conditions:[map[lastProbeTime:<nil> lastTransitionTime:2024-06-28T11:46:51Z status:False type:PodReadyToStartContainers] map[lastProbeTime:<nil> lastTransitionTime:2024-06-28T11:46:42Z reason:PodCompleted status:True type:Initialized] map[lastProbeTime:<nil> lastTransitionTime:2024-06-28T11:46:48Z reason:PodCompleted status:False type:Ready] map[lastProbeTime:<nil> lastTransitionTime:2024-06-28T11:46:48Z reason:PodCompleted status:False type:ContainersReady] map[lastProbeTime:<nil> lastTransitionTime:2024-06-28T11:46:42Z status:True type:PodScheduled]] containerStatuses:[map[containerID:containerd://7451597e9b3a1f9705ab6c87e3de2534f19ac89d21066f1c971b32b768932988 image:docker.io/howardjohn/shell:latest imageID:docker.io/howardjohn/shell@sha256:23546d7e285397394c46f44ecb503309096a24724cd1b20b841dfb49d8d6fb6f lastState:map[] name:pi ready:false restartCount:0 started:false state:map[terminated:map[containerID:containerd://7451597e9b3a1f9705ab6c87e3de2534f19ac89d21066f1c971b32b768932988 exitCode:0 finishedAt:2024-06-28T11:46:47Z reason:Completed startedAt:2024-06-28T11:46:45Z]] volumeMounts:[map[mountPath:/var/run/secrets/kubernetes.io/serviceaccount name:kube-api-access-5nnd6 readOnly:true recursiveReadOnly:Disabled]]]] hostIP:192.168.8.2 hostIPs:[map[ip:192.168.8.2]] phase:Running qosClass:BestEffort startTime:2024-06-28T11:46:42Z]
2024-06-28 13:46:52.689228 MODIFIED   pod-pi-f4v4d   <none>       Succeeded   map[conditions:[map[lastProbeTime:<nil> lastTransitionTime:2024-06-28T11:46:51Z status:False type:PodReadyToStartContainers] map[lastProbeTime:<nil> lastTransitionTime:2024-06-28T11:46:42Z reason:PodCompleted status:True type:Initialized] map[lastProbeTime:<nil> lastTransitionTime:2024-06-28T11:46:48Z reason:PodCompleted status:False type:Ready] map[lastProbeTime:<nil> lastTransitionTime:2024-06-28T11:46:48Z reason:PodCompleted status:False type:ContainersReady] map[lastProbeTime:<nil> lastTransitionTime:2024-06-28T11:46:42Z status:True type:PodScheduled]] containerStatuses:[map[containerID:containerd://7451597e9b3a1f9705ab6c87e3de2534f19ac89d21066f1c971b32b768932988 image:docker.io/howardjohn/shell:latest imageID:docker.io/howardjohn/shell@sha256:23546d7e285397394c46f44ecb503309096a24724cd1b20b841dfb49d8d6fb6f lastState:map[] name:pi ready:false restartCount:0 started:false state:map[terminated:map[containerID:containerd://7451597e9b3a1f9705ab6c87e3de2534f19ac89d21066f1c971b32b768932988 exitCode:0 finishedAt:2024-06-28T11:46:47Z reason:Completed startedAt:2024-06-28T11:46:45Z]] volumeMounts:[map[mountPath:/var/run/secrets/kubernetes.io/serviceaccount name:kube-api-access-5nnd6 readOnly:true recursiveReadOnly:Disabled]]]] hostIP:192.168.8.2 hostIPs:[map[ip:192.168.8.2]] phase:Succeeded qosClass:BestEffort startTime:2024-06-28T11:46:42Z]
2024-06-28 13:46:53.700212 MODIFIED   pod-pi-f4v4d   <none>       Succeeded   map[conditions:[map[lastProbeTime:<nil> lastTransitionTime:2024-06-28T11:46:51Z status:False type:PodReadyToStartContainers] map[lastProbeTime:<nil> lastTransitionTime:2024-06-28T11:46:42Z reason:PodCompleted status:True type:Initialized] map[lastProbeTime:<nil> lastTransitionTime:2024-06-28T11:46:48Z reason:PodCompleted status:False type:Ready] map[lastProbeTime:<nil> lastTransitionTime:2024-06-28T11:46:48Z reason:PodCompleted status:False type:ContainersReady] map[lastProbeTime:<nil> lastTransitionTime:2024-06-28T11:46:42Z status:True type:PodScheduled]] containerStatuses:[map[containerID:containerd://7451597e9b3a1f9705ab6c87e3de2534f19ac89d21066f1c971b32b768932988 image:docker.io/howardjohn/shell:latest imageID:docker.io/howardjohn/shell@sha256:23546d7e285397394c46f44ecb503309096a24724cd1b20b841dfb49d8d6fb6f lastState:map[] name:pi ready:false restartCount:0 started:false state:map[terminated:map[containerID:containerd://7451597e9b3a1f9705ab6c87e3de2534f19ac89d21066f1c971b32b768932988 exitCode:0 finishedAt:2024-06-28T11:46:47Z reason:Completed startedAt:2024-06-28T11:46:45Z]] volumeMounts:[map[mountPath:/var/run/secrets/kubernetes.io/serviceaccount name:kube-api-access-5nnd6 readOnly:true recursiveReadOnly:Disabled]]]] hostIP:192.168.8.2 hostIPs:[map[ip:192.168.8.2]] phase:Succeeded qosClass:BestEffort startTime:2024-06-28T11:46:42Z]
```
it would work if we remove the check that `newPodStatus is terminated` but it might have negative consequences in other scenarios (not sure though).

Reply:
> When is the podIP "released" by the Kubelet?
> Is it when the sandbox is stopped (
> 
https://github.com/kubernetes/kubernetes/blob/9895910846eeccf0694c577ac439f492ae2dfc3c/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L1374-L1375
> )? Or is it later?

That is correct, I tested it by calling `m.runtimeService.PodSandboxStatus(ctx, podSandbox.ID.ID, false)` just before and after `m.runtimeService.StopPodSandbox` in the link. It contains the IP before, at least on containerd.

We also have this guard [here](https://github.com/kubernetes/kubernetes/blob/e832b7023066a6f386634ad5d99e574281e1e6b4/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L1460) to only lookup IP is the standbox is ready, but at least from my testing it is redundant.

Reply:
Ok.  I would prefer this happen inside a specifically named function that generically preserves data from before the pod is stopped to keep the clarity of the method high.

Reply:
> Ok. I would prefer this happen inside a specifically named function that generically preserves data from before the pod is stopped to keep the clarity of the method high.

Sure, but let me clarify to make sure we are on the same page, are you suggesting to wrap just this line in a function `stoppedPodStatus.IPs = podStatus.IPs` like `restoreDataForSoppedPod`? or maybe it is better to wrap [this](https://github.com/kubernetes/kubernetes/blob/338edd427d1ff27696bfd505ee7e743854ac54f5/pkg/kubelet/kubelet.go#L2070-L2082) in `getStatusForStoppedContainer`?

Reply:
explictly naming a function that preserves a field would be best from my perspective

Reply:
also please add a comment to that function suggesting that kubelet restart may cause to lose those fields.

Reply:
Sure, PTAL: 66eccac7753cf08af0af1436c7f3a75048f95912.

Reply:
@smarterclayton @SergeyKanzhelev let me know if this addresses the comment: https://github.com/kubernetes/kubernetes/pull/125404#discussion_r1670710478